commit 36d096a2662c52cb04f68b7f74418c75f663cce8
Author: Your Name <you@example.com>
Date:   Tue Aug 29 15:41:57 2017 +0800

    add APP/ACTIVITY layer code
    compile:mmm ./packages/experimental/Tiny4412_led_demo/

diff --git a/packages/experimental/Tiny4412_led_demo/.classpath b/packages/experimental/Tiny4412_led_demo/.classpath
new file mode 100755
index 0000000..26bdfa6
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/.classpath
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry kind="src" path="src"/>
+	<classpathentry kind="src" path="gen"/>
+	<classpathentry kind="con" path="com.android.ide.eclipse.adt.ANDROID_FRAMEWORK"/>
+	<classpathentry exported="true" kind="con" path="com.android.ide.eclipse.adt.LIBRARIES"/>
+	<classpathentry exported="true" kind="con" path="com.android.ide.eclipse.adt.DEPENDENCIES"/>
+	<classpathentry kind="output" path="bin/classes"/>
+</classpath>
diff --git a/packages/experimental/Tiny4412_led_demo/.project b/packages/experimental/Tiny4412_led_demo/.project
new file mode 100755
index 0000000..2af3ad6
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/.project
@@ -0,0 +1,33 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<projectDescription>
+	<name>Tiny4412_led_demo</name>
+	<comment></comment>
+	<projects>
+	</projects>
+	<buildSpec>
+		<buildCommand>
+			<name>com.android.ide.eclipse.adt.ResourceManagerBuilder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
+		<buildCommand>
+			<name>com.android.ide.eclipse.adt.PreCompilerBuilder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
+		<buildCommand>
+			<name>org.eclipse.jdt.core.javabuilder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
+		<buildCommand>
+			<name>com.android.ide.eclipse.adt.ApkBuilder</name>
+			<arguments>
+			</arguments>
+		</buildCommand>
+	</buildSpec>
+	<natures>
+		<nature>com.android.ide.eclipse.adt.AndroidNature</nature>
+		<nature>org.eclipse.jdt.core.javanature</nature>
+	</natures>
+</projectDescription>
diff --git a/packages/experimental/Tiny4412_led_demo/Android.mk b/packages/experimental/Tiny4412_led_demo/Android.mk
new file mode 100755
index 0000000..18971c2
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/Android.mk
@@ -0,0 +1,12 @@
+LOCAL_PATH:= $(call my-dir)
+include $(CLEAR_VARS)
+
+LOCAL_MODULE_TAGS := optional
+
+# Only compile source java files in this apk.
+LOCAL_SRC_FILES := $(call all-java-files-under, src)
+
+LOCAL_PACKAGE_NAME := Led_demo
+
+include $(BUILD_PACKAGE)
+
diff --git a/packages/experimental/Tiny4412_led_demo/AndroidManifest.xml b/packages/experimental/Tiny4412_led_demo/AndroidManifest.xml
new file mode 100755
index 0000000..4748418
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/AndroidManifest.xml
@@ -0,0 +1,23 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    package="com.example.tiny4412_led_demo"
+    android:versionCode="1"
+    android:versionName="1.0" >
+
+    <application
+        android:allowBackup="true"
+        android:icon="@drawable/ic_launcher"
+        android:label="@string/app_name"
+        android:theme="@style/AppTheme" >
+        <activity
+            android:name="com.example.tiny4412_led_demo.MainActivity"
+            android:label="@string/app_name" >
+            <intent-filter>
+                <action android:name="android.intent.action.MAIN" />
+
+                <category android:name="android.intent.category.LAUNCHER" />
+            </intent-filter>
+        </activity>
+    </application>
+
+</manifest>
diff --git a/packages/experimental/Tiny4412_led_demo/bin/AndroidManifest.xml b/packages/experimental/Tiny4412_led_demo/bin/AndroidManifest.xml
new file mode 100755
index 0000000..ca78986
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/bin/AndroidManifest.xml
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    package="com.example.tiny4412_led_demo"
+    android:versionCode="1"
+    android:versionName="1.0" >
+
+    <uses-sdk
+        android:minSdkVersion="8"
+        android:targetSdkVersion="17" />
+
+    <application
+        android:allowBackup="true"
+        android:icon="@drawable/ic_launcher"
+        android:label="@string/app_name"
+        android:theme="@style/AppTheme" >
+        <activity
+            android:name="com.example.tiny4412_led_demo.MainActivity"
+            android:label="@string/app_name" >
+            <intent-filter>
+                <action android:name="android.intent.action.MAIN" />
+
+                <category android:name="android.intent.category.LAUNCHER" />
+            </intent-filter>
+        </activity>
+    </application>
+
+</manifest>
diff --git a/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/BuildConfig.class b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/BuildConfig.class
new file mode 100755
index 0000000..eab35d9
Binary files /dev/null and b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/BuildConfig.class differ
diff --git a/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/MainActivity.class b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/MainActivity.class
new file mode 100755
index 0000000..4a23c8b
Binary files /dev/null and b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/MainActivity.class differ
diff --git a/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$attr.class b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$attr.class
new file mode 100755
index 0000000..3b7de75
Binary files /dev/null and b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$attr.class differ
diff --git a/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$dimen.class b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$dimen.class
new file mode 100755
index 0000000..ca225f0
Binary files /dev/null and b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$dimen.class differ
diff --git a/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$drawable.class b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$drawable.class
new file mode 100755
index 0000000..3e1ba00
Binary files /dev/null and b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$drawable.class differ
diff --git a/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$id.class b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$id.class
new file mode 100755
index 0000000..489e71b
Binary files /dev/null and b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$id.class differ
diff --git a/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$layout.class b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$layout.class
new file mode 100755
index 0000000..0ee369a
Binary files /dev/null and b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$layout.class differ
diff --git a/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$menu.class b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$menu.class
new file mode 100755
index 0000000..034cd7c
Binary files /dev/null and b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$menu.class differ
diff --git a/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$string.class b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$string.class
new file mode 100755
index 0000000..c2413d9
Binary files /dev/null and b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$string.class differ
diff --git a/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$style.class b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$style.class
new file mode 100755
index 0000000..067fbf4
Binary files /dev/null and b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R$style.class differ
diff --git a/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R.class b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R.class
new file mode 100755
index 0000000..8264da7
Binary files /dev/null and b/packages/experimental/Tiny4412_led_demo/bin/classes/com/example/tiny4412_led_demo/R.class differ
diff --git a/packages/experimental/Tiny4412_led_demo/gen/com/example/tiny4412_led_demo/BuildConfig.java b/packages/experimental/Tiny4412_led_demo/gen/com/example/tiny4412_led_demo/BuildConfig.java
new file mode 100755
index 0000000..7f8a8b6
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/gen/com/example/tiny4412_led_demo/BuildConfig.java
@@ -0,0 +1,6 @@
+/** Automatically generated file. DO NOT MODIFY */
+package com.example.tiny4412_led_demo;
+
+public final class BuildConfig {
+    public final static boolean DEBUG = true;
+}
\ No newline at end of file
diff --git a/packages/experimental/Tiny4412_led_demo/gen/com/example/tiny4412_led_demo/R.java b/packages/experimental/Tiny4412_led_demo/gen/com/example/tiny4412_led_demo/R.java
new file mode 100755
index 0000000..3b8ebd2
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/gen/com/example/tiny4412_led_demo/R.java
@@ -0,0 +1,86 @@
+/* AUTO-GENERATED FILE.  DO NOT MODIFY.
+ *
+ * This class was automatically generated by the
+ * aapt tool from the resource data it found.  It
+ * should not be modified by hand.
+ */
+
+package com.example.tiny4412_led_demo;
+
+public final class R {
+    public static final class attr {
+    }
+    public static final class dimen {
+        /**  Default screen margins, per the Android Design guidelines. 
+
+         Customize dimensions originally defined in res/values/dimens.xml (such as
+         screen margins) for sw720dp devices (e.g. 10" tablets) in landscape here.
+    
+         */
+        public static final int activity_horizontal_margin=0x7f040000;
+        public static final int activity_vertical_margin=0x7f040001;
+    }
+    public static final class drawable {
+        public static final int ic_launcher=0x7f020000;
+    }
+    public static final class id {
+        public static final int action_settings=0x7f08000c;
+        public static final int bt_off=0x7f080002;
+        public static final int bt_off2=0x7f080005;
+        public static final int bt_off3=0x7f080008;
+        public static final int bt_off4=0x7f08000b;
+        public static final int bt_on=0x7f080001;
+        public static final int bt_on2=0x7f080004;
+        public static final int bt_on3=0x7f080007;
+        public static final int bt_on4=0x7f08000a;
+        public static final int tv1=0x7f080000;
+        public static final int tv2=0x7f080003;
+        public static final int tv3=0x7f080006;
+        public static final int tv4=0x7f080009;
+    }
+    public static final class layout {
+        public static final int activity_main=0x7f030000;
+    }
+    public static final class menu {
+        public static final int main=0x7f070000;
+    }
+    public static final class string {
+        public static final int OFF=0x7f050008;
+        public static final int ON=0x7f050007;
+        public static final int action_settings=0x7f050001;
+        public static final int app_name=0x7f050000;
+        public static final int hello_world=0x7f050002;
+        public static final int led1=0x7f050003;
+        public static final int led2=0x7f050004;
+        public static final int led3=0x7f050005;
+        public static final int led4=0x7f050006;
+    }
+    public static final class style {
+        /** 
+        Base application theme, dependent on API level. This theme is replaced
+        by AppBaseTheme from res/values-vXX/styles.xml on newer devices.
+    
+
+            Theme customizations available in newer API levels can go in
+            res/values-vXX/styles.xml, while customizations related to
+            backward-compatibility can go here.
+        
+
+        Base application theme for API 11+. This theme completely replaces
+        AppBaseTheme from res/values/styles.xml on API 11+ devices.
+    
+ API 11 theme customizations can go here. 
+
+        Base application theme for API 14+. This theme completely replaces
+        AppBaseTheme from BOTH res/values/styles.xml and
+        res/values-v11/styles.xml on API 14+ devices.
+    
+ API 14 theme customizations can go here. 
+         */
+        public static final int AppBaseTheme=0x7f060000;
+        /**  Application theme. 
+ All customizations that are NOT specific to a particular API-level can go here. 
+         */
+        public static final int AppTheme=0x7f060001;
+    }
+}
diff --git a/packages/experimental/Tiny4412_led_demo/ic_launcher-web.png b/packages/experimental/Tiny4412_led_demo/ic_launcher-web.png
new file mode 100755
index 0000000..a18cbb4
Binary files /dev/null and b/packages/experimental/Tiny4412_led_demo/ic_launcher-web.png differ
diff --git a/packages/experimental/Tiny4412_led_demo/libs/android-support-v4.jar b/packages/experimental/Tiny4412_led_demo/libs/android-support-v4.jar
new file mode 100755
index 0000000..96644ed
Binary files /dev/null and b/packages/experimental/Tiny4412_led_demo/libs/android-support-v4.jar differ
diff --git a/packages/experimental/Tiny4412_led_demo/proguard-project.txt b/packages/experimental/Tiny4412_led_demo/proguard-project.txt
new file mode 100755
index 0000000..f2fe155
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/proguard-project.txt
@@ -0,0 +1,20 @@
+# To enable ProGuard in your project, edit project.properties
+# to define the proguard.config property as described in that file.
+#
+# Add project specific ProGuard rules here.
+# By default, the flags in this file are appended to flags specified
+# in ${sdk.dir}/tools/proguard/proguard-android.txt
+# You can edit the include path and order by changing the ProGuard
+# include property in project.properties.
+#
+# For more details, see
+#   http://developer.android.com/guide/developing/tools/proguard.html
+
+# Add any project specific keep options here:
+
+# If your project uses WebView with JS, uncomment the following
+# and specify the fully qualified class name to the JavaScript interface
+# class:
+#-keepclassmembers class fqcn.of.javascript.interface.for.webview {
+#   public *;
+#}
diff --git a/packages/experimental/Tiny4412_led_demo/project.properties b/packages/experimental/Tiny4412_led_demo/project.properties
new file mode 100755
index 0000000..4ab1256
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/project.properties
@@ -0,0 +1,14 @@
+# This file is automatically generated by Android Tools.
+# Do not modify this file -- YOUR CHANGES WILL BE ERASED!
+#
+# This file must be checked in Version Control Systems.
+#
+# To customize properties used by the Ant build system edit
+# "ant.properties", and override values to adapt the script to your
+# project structure.
+#
+# To enable ProGuard to shrink and obfuscate your code, uncomment this (available properties: sdk.dir, user.home):
+#proguard.config=${sdk.dir}/tools/proguard/proguard-android.txt:proguard-project.txt
+
+# Project target.
+target=android-19
diff --git a/packages/experimental/Tiny4412_led_demo/res/drawable-hdpi/ic_launcher.png b/packages/experimental/Tiny4412_led_demo/res/drawable-hdpi/ic_launcher.png
new file mode 100755
index 0000000..288b665
Binary files /dev/null and b/packages/experimental/Tiny4412_led_demo/res/drawable-hdpi/ic_launcher.png differ
diff --git a/packages/experimental/Tiny4412_led_demo/res/drawable-mdpi/ic_launcher.png b/packages/experimental/Tiny4412_led_demo/res/drawable-mdpi/ic_launcher.png
new file mode 100755
index 0000000..6ae570b
Binary files /dev/null and b/packages/experimental/Tiny4412_led_demo/res/drawable-mdpi/ic_launcher.png differ
diff --git a/packages/experimental/Tiny4412_led_demo/res/drawable-xhdpi/ic_launcher.png b/packages/experimental/Tiny4412_led_demo/res/drawable-xhdpi/ic_launcher.png
new file mode 100755
index 0000000..d4fb7cd
Binary files /dev/null and b/packages/experimental/Tiny4412_led_demo/res/drawable-xhdpi/ic_launcher.png differ
diff --git a/packages/experimental/Tiny4412_led_demo/res/layout/activity_main.xml b/packages/experimental/Tiny4412_led_demo/res/layout/activity_main.xml
new file mode 100755
index 0000000..8ff2607
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/res/layout/activity_main.xml
@@ -0,0 +1,107 @@
+<LinearLayout xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    android:layout_width="match_parent"
+    android:layout_height="match_parent"
+    android:orientation="vertical" >
+
+    <LinearLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:gravity="center"
+        android:orientation="horizontal" >
+
+        <TextView
+            android:id="@+id/tv1"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/led1" />
+
+        <Button
+            android:id="@+id/bt_on"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/ON" />
+
+        <Button
+            android:id="@+id/bt_off"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/OFF" />
+    </LinearLayout>
+
+    <LinearLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:gravity="center"
+        android:orientation="horizontal" >
+
+        <TextView
+            android:id="@+id/tv2"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/led2" />
+
+        <Button
+            android:id="@+id/bt_on2"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/ON" />
+
+        <Button
+            android:id="@+id/bt_off2"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/OFF" />
+    </LinearLayout>
+
+    <LinearLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:gravity="center"
+        android:orientation="horizontal" >
+
+        <TextView
+            android:id="@+id/tv3"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/led3" />
+
+        <Button
+            android:id="@+id/bt_on3"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/ON" />
+
+        <Button
+            android:id="@+id/bt_off3"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/OFF" />
+    </LinearLayout>
+
+    <LinearLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:gravity="center"
+        android:orientation="horizontal" >
+
+        <TextView
+            android:id="@+id/tv4"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/led4" />
+
+        <Button
+            android:id="@+id/bt_on4"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/ON" />
+
+        <Button
+            android:id="@+id/bt_off4"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@string/OFF" />
+    </LinearLayout>
+
+</LinearLayout>
\ No newline at end of file
diff --git a/packages/experimental/Tiny4412_led_demo/res/menu/main.xml b/packages/experimental/Tiny4412_led_demo/res/menu/main.xml
new file mode 100755
index 0000000..d122a4b
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/res/menu/main.xml
@@ -0,0 +1,9 @@
+<menu xmlns:android="http://schemas.android.com/apk/res/android" >
+
+    <item
+        android:id="@+id/action_settings"
+        android:orderInCategory="100"
+        android:showAsAction="never"
+        android:title="@string/action_settings"/>
+
+</menu>
diff --git a/packages/experimental/Tiny4412_led_demo/res/values-sw600dp/dimens.xml b/packages/experimental/Tiny4412_led_demo/res/values-sw600dp/dimens.xml
new file mode 100755
index 0000000..c876987
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/res/values-sw600dp/dimens.xml
@@ -0,0 +1,8 @@
+<resources>
+
+    <!--
+         Customize dimensions originally defined in res/values/dimens.xml (such as
+         screen margins) for sw600dp devices (e.g. 7" tablets) here.
+    -->
+
+</resources>
diff --git a/packages/experimental/Tiny4412_led_demo/res/values-sw720dp-land/dimens.xml b/packages/experimental/Tiny4412_led_demo/res/values-sw720dp-land/dimens.xml
new file mode 100755
index 0000000..0df3067
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/res/values-sw720dp-land/dimens.xml
@@ -0,0 +1,9 @@
+<resources>
+
+    <!--
+         Customize dimensions originally defined in res/values/dimens.xml (such as
+         screen margins) for sw720dp devices (e.g. 10" tablets) in landscape here.
+    -->
+    <dimen name="activity_horizontal_margin">128dp</dimen>
+
+</resources>
diff --git a/packages/experimental/Tiny4412_led_demo/res/values-v11/styles.xml b/packages/experimental/Tiny4412_led_demo/res/values-v11/styles.xml
new file mode 100755
index 0000000..e3ef53d
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/res/values-v11/styles.xml
@@ -0,0 +1,11 @@
+<resources>
+
+    <!--
+        Base application theme for API 11+. This theme completely replaces
+        AppBaseTheme from res/values/styles.xml on API 11+ devices.
+    -->
+    <style name="AppBaseTheme" parent="android:Theme.Holo.Light">
+        <!-- API 11 theme customizations can go here. -->
+    </style>
+
+</resources>
diff --git a/packages/experimental/Tiny4412_led_demo/res/values-v14/styles.xml b/packages/experimental/Tiny4412_led_demo/res/values-v14/styles.xml
new file mode 100755
index 0000000..94dd245
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/res/values-v14/styles.xml
@@ -0,0 +1,12 @@
+<resources>
+
+    <!--
+        Base application theme for API 14+. This theme completely replaces
+        AppBaseTheme from BOTH res/values/styles.xml and
+        res/values-v11/styles.xml on API 14+ devices.
+    -->
+    <style name="AppBaseTheme" parent="android:Theme.Holo.Light.DarkActionBar">
+        <!-- API 14 theme customizations can go here. -->
+    </style>
+
+</resources>
diff --git a/packages/experimental/Tiny4412_led_demo/res/values/dimens.xml b/packages/experimental/Tiny4412_led_demo/res/values/dimens.xml
new file mode 100755
index 0000000..2e0e2ae
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/res/values/dimens.xml
@@ -0,0 +1,7 @@
+<resources>
+
+    <!-- Default screen margins, per the Android Design guidelines. -->
+    <dimen name="activity_horizontal_margin">16dp</dimen>
+    <dimen name="activity_vertical_margin">16dp</dimen>
+
+</resources>
diff --git a/packages/experimental/Tiny4412_led_demo/res/values/strings.xml b/packages/experimental/Tiny4412_led_demo/res/values/strings.xml
new file mode 100755
index 0000000..5e71b48
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/res/values/strings.xml
@@ -0,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+
+    <string name="app_name">Tiny4412_led_demo</string>
+    <string name="action_settings">Settings</string>
+    <string name="hello_world">Hello world!</string>
+    <string name="led1">LED1</string>
+    <string name="led2">LED2</string>
+    <string name="led3">LED3</string>
+    <string name="led4">LED4</string>
+    <string name="ON">ON</string>
+    <string name="OFF">OFF</string>
+
+</resources>
diff --git a/packages/experimental/Tiny4412_led_demo/res/values/styles.xml b/packages/experimental/Tiny4412_led_demo/res/values/styles.xml
new file mode 100755
index 0000000..4ea9326
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/res/values/styles.xml
@@ -0,0 +1,20 @@
+<resources>
+
+    <!--
+        Base application theme, dependent on API level. This theme is replaced
+        by AppBaseTheme from res/values-vXX/styles.xml on newer devices.
+    -->
+    <style name="AppBaseTheme" parent="android:Theme.Light">
+        <!--
+            Theme customizations available in newer API levels can go in
+            res/values-vXX/styles.xml, while customizations related to
+            backward-compatibility can go here.
+        -->
+    </style>
+
+    <!-- Application theme. -->
+    <style name="AppTheme" parent="AppBaseTheme">
+        <!-- All customizations that are NOT specific to a particular API-level can go here. -->
+    </style>
+
+</resources>
diff --git a/packages/experimental/Tiny4412_led_demo/src/com/example/tiny4412_led_demo/MainActivity.java b/packages/experimental/Tiny4412_led_demo/src/com/example/tiny4412_led_demo/MainActivity.java
new file mode 100755
index 0000000..73b965e
--- /dev/null
+++ b/packages/experimental/Tiny4412_led_demo/src/com/example/tiny4412_led_demo/MainActivity.java
@@ -0,0 +1,156 @@
+package com.example.tiny4412_led_demo;
+
+import android.os.ILed_demo_service;
+import android.os.RemoteException;
+import android.os.ServiceManager;
+import android.os.Bundle;
+import android.app.Activity;
+import android.util.Log;
+import android.view.Menu;
+import android.view.View;
+import android.view.View.OnClickListener;
+import android.widget.Button;
+
+public class MainActivity extends Activity implements OnClickListener
+{
+    private final static String LOG_TAG = "com.example.tiny4412_led_demo.MainActivity";
+
+    private ILed_demo_service  Led_demo_service = null;
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+
+        Led_demo_service = ILed_demo_service.Stub.asInterface(ServiceManager.getService("led_demo"));
+
+        Button bt_on = (Button)findViewById(R.id.bt_on);
+        Button bt_off = (Button)findViewById(R.id.bt_off);
+
+        bt_on.setOnClickListener(this);
+        bt_off.setOnClickListener(this);
+
+		Button bt_on2 = (Button)findViewById(R.id.bt_on2);
+		Button bt_off2 = (Button)findViewById(R.id.bt_off2);
+		
+		bt_on2.setOnClickListener(this);
+		bt_off2.setOnClickListener(this);
+		
+		Button bt_on3 = (Button)findViewById(R.id.bt_on3);
+		Button bt_off3 = (Button)findViewById(R.id.bt_off3);
+		
+		bt_on3.setOnClickListener(this);
+		bt_off3.setOnClickListener(this);
+		
+		Button bt_on4 = (Button)findViewById(R.id.bt_on4);
+		Button bt_off4 = (Button)findViewById(R.id.bt_off4);
+		
+		bt_on4.setOnClickListener(this);
+		bt_off4.setOnClickListener(this);
+		
+		Log.i(LOG_TAG, "MainActivity created.");
+	}
+
+    @Override
+    public boolean onCreateOptionsMenu(Menu menu) {
+        // Inflate the menu; this adds items to the action bar if it is present.
+        getMenuInflater().inflate(R.menu.main, menu);
+        return true;
+    }
+
+    @Override
+    public void onClick(View v) {
+        // TODO Auto-generated method stub
+        switch(v.getId())
+        {
+            case R.id.bt_on:
+                try
+                {
+                    Led_demo_service.led_set_ON(1);
+                    Log.e(LOG_TAG, "set led on is execute.");
+                    break;
+                } catch (RemoteException e)
+                {
+                    Log.e(LOG_TAG, "Remote Exception while set led on.");
+                }
+
+            case R.id.bt_off:
+                try
+                {
+                    Led_demo_service.led_set_OFF(1);
+                    Log.e(LOG_TAG, "set led off is execute.");
+                    break;
+                } catch (RemoteException e)
+                {
+                    Log.e(LOG_TAG, "Remote Exception while set led off.");
+                }
+
+            case R.id.bt_on2:
+                try
+                {
+                    Led_demo_service.led_set_ON(2);
+                    Log.e(LOG_TAG, "set led on is execute.");
+                    break;
+                } catch (RemoteException e)
+                {
+                    Log.e(LOG_TAG, "Remote Exception while set led on.");
+                }
+
+            case R.id.bt_off2:
+                try
+                {
+                    Led_demo_service.led_set_OFF(2);
+                    Log.e(LOG_TAG, "set led off is execute.");
+                    break;
+                } catch (RemoteException e)
+                {
+                    Log.e(LOG_TAG, "Remote Exception while set led off.");
+                }
+
+            case R.id.bt_on3:
+                try
+                {
+                    Led_demo_service.led_set_ON(3);
+                    Log.e(LOG_TAG, "set led on is execute.");
+                    break;
+                } catch (RemoteException e)
+                {
+                    Log.e(LOG_TAG, "Remote Exception while set led on.");
+                }
+
+            case R.id.bt_off3:
+                try
+                {
+                    Led_demo_service.led_set_OFF(3);
+                    Log.e(LOG_TAG, "set led off is execute.");
+                    break;
+                } catch (RemoteException e)
+                {
+                    Log.e(LOG_TAG, "Remote Exception while set led off.");
+                }
+
+            case R.id.bt_on4:
+                try
+                {
+                    Led_demo_service.led_set_ON(4);
+                    Log.e(LOG_TAG, "set led on is execute.");
+                    break;
+                } catch (RemoteException e)
+                {
+                    Log.e(LOG_TAG, "Remote Exception while set led on.");
+                }
+
+            case R.id.bt_off4:
+                try
+                {
+                    Led_demo_service.led_set_OFF(4);
+                    Log.e(LOG_TAG, "set led off is execute.");
+                    break;
+                } catch (RemoteException e)
+                {
+                    Log.e(LOG_TAG, "Remote Exception while set led off.");
+                }
+        }
+
+    }
+
+}
